app/src/main/java/net/squanchy/support/widget/CircleImageView.kt

```diff
@@ -0,0 +1,241 @@
+        defStyleAttr: Int = 0
+) : ImageViewWithForeground(context, attrs, defStyleAttr) {
+
+    private var necessaireOptional: Necessaire? = null
+
+    private val isZeroSize: Boolean
+        get() = width == 0 || height == 0
+
+    init {
+        super.setOutlineProvider(CircularOutlineProvider)
+        super.setClipToOutline(true)
+        super.setScaleType(ImageView.ScaleType.CENTER_CROP)
+    }
+
+    override fun setOutlineProvider(provider: ViewOutlineProvider): Nothing {
+        throw UnsupportedOperationException("Cannot set an outline provider on a CircleImageView")
+    }
+
+    override fun setClipToOutline(clipToOutline: Boolean): Nothing {
+        throw UnsupportedOperationException("Cannot set clipping to outline on a CircleImageView")
+    }
+
+    override fun setScaleType(scaleType: ImageView.ScaleType): Nothing {
+        throw UnsupportedOperationException("Cannot set scale type on a CircleImageView")
+    }
+
+    override fun getOutlineProvider(): ViewOutlineProvider = CircularOutlineProvider
+
+    override fun setImageBitmap(bm: Bitmap?) {
+        super.setImageBitmap(bm)
+        updateFor(drawable)
+    }
+
+    override fun setImageDrawable(drawable: Drawable?) {
+        super.setImageDrawable(drawable)
+        updateFor(drawable)
+    }
+
+    override fun setImageResource(@DrawableRes resId: Int) {
+        super.setImageResource(resId)
+        updateFor(drawable)
+    }
+
+    override fun setImageURI(uri: Uri?) {
+        super.setImageURI(uri)
+        updateFor(drawable)
+    }
+
+    private fun updateFor(drawable: Drawable?) {
+        val bitmap = extractBitmapFrom(drawable)
+        necessaireOptional = necessaireFor(bitmap)
+        invalidate()
+    }
+
+    private fun extractBitmapFrom(drawable: Drawable?): Bitmap? {
+        return when (drawable) {
+            null -> null
+            is BitmapDrawable -> (drawable as? BitmapDrawable)?.bitmap
+            else -> tryCreatingBitmapFrom(drawable)
+        }
+    }
+
+    private fun tryCreatingBitmapFrom(drawable: Drawable): Bitmap? {
+        return try {
+            val bitmap = createBitmapFor(drawable)
+            val canvas = Canvas(bitmap)
+            drawable.setBounds(0, 0, canvas.width, canvas.height)
+            drawable.draw(canvas)
+            bitmap
+        } catch (e: IllegalArgumentException) {
+            Timber.e(e)
+            null
+        }
+    }
+
+    private fun createBitmapFor(drawable: Drawable): Bitmap {
+        return if (drawable is ColorDrawable) {
+            Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)
+        } else {
+            Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
+        }
+    }
+
+    private fun necessaireFor(bitmapOptional: Bitmap?): Necessaire? {
+        return if (isZeroSize) null else bitmapOptional?.let { toNecessaireOptional(it) }
+    }
+
+    private fun toNecessaireOptional(bitmap: Bitmap): Necessaire {
```

Comment:
This should be an extension function: `private fun Bitmap.toNecessaire(): Necessaire` (also it returns a non-null value so it's by no means "optional"